File: DEMO ALAN/backend-node/compatibility.js
import { readFile } from "fs/promises";

let db = null;

export async function loadDB(filePath) {
  const raw = await readFile(filePath, "utf-8");
  db = JSON.parse(raw);
}

export function parseInsert(code) {
  code = code.toUpperCase().replace(/[\s\-]/g, "");
  const result = { raw: code, shapeLetter: null, shapeInfo: null, isValid: false };
  if (code.length < 4) return result;
  if (db.shapes[code[0]]) {
    result.shapeLetter = code[0];
    result.shapeInfo = db.shapes[code[0]];
    result.isValid = true;
  }
  const numMatch = code.match(/(\d{4,6})/);
  result.sizeCode = numMatch ? numMatch[1] : null;
  result.clearanceLetter = code[1] && db.clearance_angles[code[1]] ? code[1] : null;
  result.clearanceInfo = result.clearanceLetter ? db.clearance_angles[result.clearanceLetter] : null;
  result.toleranceLetter = code[2] && db.tolerance_classes[code[2]] ? code[2] : null;
  result.toleranceInfo = result.toleranceLetter ? db.tolerance_classes[result.toleranceLetter] : null;
  return result;
}

export function getCompatibility(insertCode) {
  const parsed = parseInsert(insertCode);
  if (!parsed.isValid) return { error: "Неверный код пластины", insert_code: insertCode };

  const shape = parsed.shapeInfo;
  const sl = parsed.shapeLetter;

  const compatHolders = [];
  for (const [bk, brand] of Object.entries(db.brands)) {
    for (const [lineName, line] of Object.entries(brand.lines)) {
      const matched = (line.inserts || []).filter(i => i.toUpperCase().startsWith(sl));
      if (matched.length) {
        compatHolders.push({
          brand: brand.name, brand_key: bk, country: brand.country,
          logo_color: brand.logo_color, line: lineName,
          line_description: line.description, holders: line.holders,
          matched_inserts: matched.slice(0, 4),
        });
      }
    }
  }

  return {
    insert_code: insertCode,
    parsed: {
      shape: `${sl} — ${shape.name}`,
      clearance: parsed.clearanceInfo,
      tolerance: parsed.toleranceInfo,
      size_code: parsed.sizeCode,
    },
    applications: shape.applications,
    industries: shape.industries,
    compatible_holders_by_brand: compatHolders,
    positive_variants: shape.positive_variants || [],
    negative_variants: shape.negative_variants || [],
    holder_prefix_patterns: shape.holder_prefixes,
  };
}

export function getHolders() {
  return Object.entries(db.brands).map(([bk, brand]) => ({
    brand_key: bk, brand: brand.name, country: brand.country,
    logo_color: brand.logo_color,
    lines: Object.entries(brand.lines).map(([name, l]) => ({
      name, description: l.description, holders: l.holders, inserts: l.inserts || [],
    })),
  }));
}

export function getInsertTypes() {
  return {
    shapes: Object.fromEntries(
      Object.entries(db.shapes).map(([k, v]) => [k, { name: v.name, description: v.description, applications: v.applications }])
    ),
    popular_inserts: db.popular_inserts,
    materials: db.materials,
  };
}

export function searchByHolder(prefix) {
  prefix = prefix.toUpperCase();
  const results = [];
  for (const [, brand] of Object.entries(db.brands)) {
    for (const [lineName, line] of Object.entries(brand.lines)) {
      const matched = (line.holders || []).filter(h => h.toUpperCase().startsWith(prefix));
      if (matched.length) {
        results.push({ brand: brand.name, line: lineName, holders: matched, compatible_inserts: line.inserts || [] });
      }
    }
  }
  return results;
}

File: DEMO ALAN/backend/compatibility.py
import json
import os
import re
from typing import Dict, List

DB_PATH = os.path.join(os.path.dirname(__file__), "data", "tools.json")


class CompatibilityDB:
    def __init__(self):
        with open(DB_PATH, encoding="utf-8") as f:
            self.db = json.load(f)

    def get_categories(self) -> Dict:
        return self.db.get("categories", {})

    def get_category_popular(self, category: str) -> List[str]:
        cat = self.db.get("categories", {}).get(category, {})
        return cat.get("popular", [])

    def parse_insert_code(self, code: str) -> Dict:
        code = code.upper().replace(" ", "").replace("-", "")
        result = {"raw": code, "shape_letter": None, "shape_info": None, "is_valid": False}
        if len(code) < 4:
            return result
        if code[0] in self.db["shapes"]:
            result["shape_letter"] = code[0]
            result["shape_info"] = self.db["shapes"][code[0]]
            result["is_valid"] = True
        if len(code) > 1 and code[1] in self.db.get("clearance_angles", {}):
            result["clearance_info"] = self.db["clearance_angles"][code[1]]
        if len(code) > 2 and code[2] in self.db.get("tolerance_classes", {}):
            result["tolerance_info"] = self.db["tolerance_classes"][code[2]]
        m = re.search(r"(\d{4,6})", code)
        result["size_code"] = m.group(1) if m else None
        return result

    def get_compatibility(self, insert_code: str) -> Dict:
        parsed = self.parse_insert_code(insert_code)
        if not parsed["is_valid"]:
            return {"error": "Неверный код пластины", "insert_code": insert_code}
        shape = parsed["shape_info"]
        sl = parsed["shape_letter"]
        compat = []
        for bk, brand in self.db["brands"].items():
            for ln, line in brand["lines"].items():
                matched = [i for i in line.get("inserts", []) if i.upper().startswith(sl)]
                if matched:
                    compat.append({
                        "brand": brand["name"], "brand_key": bk,
                        "country": brand["country"], "logo_color": brand["logo_color"],
                        "line": ln, "line_description": line["description"],
                        "holders": line["holders"], "matched_inserts": matched[:4],
                    })
        return {
            "insert_code": insert_code,
            "parsed": {
                "shape": f"{sl} — {shape['name']}",
                "clearance": parsed.get("clearance_info"),
                "tolerance": parsed.get("tolerance_info"),
                "size_code": parsed.get("size_code"),
            },
            "applications": shape["applications"],
            "industries": shape["industries"],
            "compatible_holders_by_brand": compat,
            "positive_variants": shape.get("positive_variants", []),
            "negative_variants": shape.get("negative_variants", []),
            "holder_prefix_patterns": shape["holder_prefixes"],
        }

    def get_holders(self) -> List[Dict]:
        return [
            {
                "brand_key": bk, "brand": b["name"],
                "country": b["country"], "logo_color": b["logo_color"],
                "lines": [
                    {"name": ln, "description": l["description"],
                     "holders": l["holders"], "inserts": l.get("inserts", [])}
                    for ln, l in b["lines"].items()
                ],
            }
            for bk, b in self.db["brands"].items()
        ]

    def get_insert_types(self) -> Dict:
        return {
            "shapes": {k: {"name": v["name"], "description": v["description"], "applications": v["applications"]}
                       for k, v in self.db["shapes"].items()},
            "popular_inserts": self.db["popular_inserts"],
            "materials": self.db["materials"],
            "categories": self.db.get("categories", {}),
        }

    def search_by_holder(self, prefix: str) -> List[Dict]:
        prefix = prefix.upper()
        results = []
        for _, brand in self.db["brands"].items():
            for ln, line in brand["lines"].items():
                matched = [h for h in line["holders"] if h.upper().startswith(prefix)]
                if matched:
                    results.append({
                        "brand": brand["name"], "line": ln,
                        "holders": matched, "compatible_inserts": line.get("inserts", []),
                    })
        return results
